man/recsel.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.10.
.TH RECSEL "1" "April 2014" "recsel 1.7" "User Commands"
.SH NAME
recsel \- print records from a recfile
.SH SYNOPSIS
.B recsel
[\fIOPTION\fR]... [\fI-t TYPE\fR] [\fI-j FIELD\fR] [\fI-n INDEXES | -e RECORD_EXPR | -q STR | -m NUM\fR] [\fI-c | (-p|-P) FIELD_EXPR\fR] [\fIFILE\fR]...
.SH DESCRIPTION
Select and print rec data.
.TP
\fB\-d\fR, \fB\-\-include\-descriptors\fR
print record descriptors along with the matched
records.
.TP
\fB\-C\fR, \fB\-\-collapse\fR
do not section the result in records with newlines.
.TP
\fB\-S\fR, \fB\-\-sort\fR=\fIFIELD\fR,...
sort the output by the specified fields.
.TP
\fB\-G\fR, \fB\-\-group\-by\fR=\fIFIELD\fR,...
group records by the specified fields.
.TP
\fB\-U\fR, \fB\-\-uniq\fR
remove duplicated fields in the output records.
.TP
\fB\-s\fR, \fB\-\-password\fR=\fISTR\fR
decrypt confidential fields with the given password.
.TP
\fB\-\-help\fR
print a help message and exit.
.TP
\fB\-\-version\fR
show version and exit.
.SS "Record selection options:"
.TP
\fB\-i\fR, \fB\-\-case\-insensitive\fR
make strings case\-insensitive in selection
expressions.
.TP
\fB\-t\fR, \fB\-\-type\fR=\fITYPE\fR
operate on records of the specified type only.
.TP
\fB\-e\fR, \fB\-\-expression\fR=\fIRECORD_EXPR\fR
selection expression.
.TP
\fB\-q\fR, \fB\-\-quick\fR=\fISTR\fR
select records with fields containing a string.
.TP
\fB\-n\fR, \fB\-\-number\fR=\fINUM\fR,...
select specific records by position, with ranges.
.TP
\fB\-m\fR, \fB\-\-random\fR=\fINUM\fR
select a given number of random records.
.TP
\fB\-j\fR, \fB\-\-join\fR=\fIFIELD\fR
perform an inner join using the specified field.
.SS "Output options:"
.TP
\fB\-p\fR, \fB\-\-print\fR=\fIFIELDS\fR
comma\-separated list of fields to print for each
matching record.
.TP
\fB\-P\fR, \fB\-\-print\-values\fR=\fIFIELDS\fR
as \fB\-p\fR, but print only the values of the selected
fields.
.TP
\fB\-R\fR, \fB\-\-print\-row\fR=\fIFIELDS\fR
as \fB\-P\fR, but separate the values with spaces instead
of newlines.
.TP
\fB\-c\fR, \fB\-\-count\fR
print a count of the matching records instead of
the records themselves.
.SS "Special options:"
.TP
\fB\-\-print\-sexps\fR
print the data in sexps instead of rec format.
.SH AUTHOR
Written by Jose E. Marchesi.
.SH "REPORTING BUGS"
Report bugs to: bug\-recutils@gnu.org
.br
GNU recutils home page: <http://www.gnu.org/software/recutils/>
.br
General help using GNU software: <http://www.gnu.org/gethelp/>
.SH COPYRIGHT
Copyright \(co 2010, 2011, 2012, 2013, 2014 Jose E. Marchesi.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B recsel
is maintained as a Texinfo manual.  If the
.B info
and
.B recsel
programs are properly installed at your site, the command
.IP
.B info recutils
.PP
should give you access to the complete manual.


NEWS
This NEWS file records noteworthy changes, very tersely.
See the manual for detailed information.

  Copyright 2010-2014 Jose E. Marchesi.

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

 Version 1.7 (25 March 2014)
* Utilities:
  . rec2csv now supports a -d option to change the delimiter
    character in CSV files.
  . It is now possible to specify several recfiles in the
    command line containing anonymous records.
* Format:
  . Support for restricting the fields that can appear in records
    with the new %allowed special field.
  . If a tab separates a field name and a field value is now
    ignored and not considered part of the field value.
* librec:
  . The current version of the library is now 1.
* Emacs mode:
  . It now possible to jump several records while navigating
    using a prefix argument to the next/previous commands.
* Documentation:
  . The examples in the manual have been improved, and well as many
    other aspects: formatting, English, etc.  Special thanks to Karl
    Berry for pointing out the problems and suggesting solutions.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.6 (03 November 2013)
* Utilities:
  . Support for joins and foreign keys.
  . Grouping records by one or more fields.
  . Sorting recors by several fields.
  . Specifying rewrite rules in field expressions.
  . Writing field values in rows.
  . Dot-notation in selection expressions.
  . Support for the >= and <= operators in selection expressions.
  . Aggregated functions: sum, min, max, avg.
  . New bash loadable builtin `readrec'.
* Format:
  . New UUID field type.
  . User provided arbitrary constraints with %constraint.
  . Pre-defined constants MIN and MAX can now be used in ranges.
* librec:
  . New high-level API: rec_db_query, rec_db_insert, rec_db_delete and
    rec_db_set.
  . Documentation in the rec.h header file extended and improved.
* Emacs mode:
  . Texinfo manual describing the usage of rec-mode.
  . Support for a mode hook.
  . Comments are now supported in records.
  . add-change-log-entry in a recfile now does the right thing.
  . Commands for copying/killing fields and records.
  . Support for undo in navigation mode.
  . Field folding.
  . Editing methods for the several field types.
  . Statistics.
* Documentation:
  . The user manual has been restructured and expanded.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.5 (13 January 2012)
* The utilities will now ask interactively for a password if it was
  not provided with the -s command line option.  This avoids security
  problems related to shell history files.
* Support for octal and hexadecimal numbers has been added.  They can
  be used in both the records and selection expressions.
* It is now possible to select a given number of random records in
  many of the utilities using the -m command line option.
* The -n option now accepts a list of indexes, supporting ranges.
* The new -U (uniq) option for recsel removes duplicated fields in the
  output records.
* The new -q (quick) option allows to quickly search for the desired
  record without having to provide a complete selection expression.
* Auto generated fields are now considered integers by default.  This
  avoids repetitive patterns in record descriptors involving %auto and
  %type.
* Tab characters are now allowed in blank lines betwwen records.
* The API in rec.h is now better documented with comments, and
  improved.
* recfix now exits with an error status if there is a parse error in
  some input file.
* The usage of the internal data structures has been _vastly_
  improved, resulting in a much faster operation.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.4 (03 November 2011)
* Support for encryption with the new %confidential special field.
* Support for named types with the new %typedef special field.
* Support for sorting with the new %sort special field.
* Recutils support for org-babel is now included in the distribution
  as ob-rec.el.  It allows to read rec data into org documents in both
  raw and table formats.
* It is now possible to use double quotes for delimiting string
  literals in selection expressions, as well as single quotes.
  Unescapted newline characters are also supported.
* The manual has been vastly improved.
* recfix now supports several operations: check, sort, encrypt and
  decrypt.
* recins now generates parseable locale independent dates when
  generating time-stamps in auto-fields.
* Lots of bug fixes.

 Version 1.3 (11 February 2011)
* A quite bad (and embarrassing) bug introduced in 1.2 has been
  fixed.  The bug was causing recins, recdel and recset to report
  random integrity failures, due to a uninitialized variable.
* New special field %size, that allows to define constraints in the
  size of record sets.
* Several fixes in the user manual.
* Little portability issue that prevented one of the tests to compile
  in macos x has been fixed.
* Translations updates.

 Version 1.2 (4 February 2011)
* New utility rec2csv, to export rec data to csv files.
* Support for auto-incremented fields and timestamps with the new
  %auto field in record descriptors.
* Support for comments in enum type descriptions.
* recset now supports a new "set or add field" operation, implementing
  the common case of setting the value of a field in case it exists,
  or else add it to the record.
* The flavor of regular expressions accepted in both selection
  expressions and type descriptions is now documented in the user
  manual.
* Translations for German, Finnish, Swedish and Dutch.
* recdel now performs an integrity check before to complete the
  deletion.
* The utilities no longer "expand" the contents of remote descriptors
  in the local ones.

utils/recsel.c
/* -*- mode: C -*-
 *
 *       File:         recsel.c
 *       Date:         Fri Jan  1 23:12:38 2010
 *
 *       GNU recutils - recsel
 *
 */

/* Copyright (C) 2010-2014 Jose E. Marchesi */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <getopt.h>
#include <string.h>
#include <stdlib.h>
#include <xalloc.h>
#include <gettext.h>
#define _(str) gettext (str)

#include <rec.h>
#include <recutl.h>

/* Forward prototypes.  */
void recsel_parse_args (int argc, char **argv);
bool recsel_process_data (rec_db_t db);

/*
 * Global variables
 */

char      *recutl_sex_str      = NULL;
rec_sex_t  recutl_sex          = NULL;
char      *recutl_quick_str    = NULL;
char      *recsel_fex_str      = NULL;
rec_fex_t  recsel_fex          = NULL;
char      *recutl_type         = NULL;
bool       recsel_collapse     = false;
bool       recsel_count        = false;
bool       recutl_insensitive  = false;
bool       recsel_descriptors  = false;
rec_fex_t  recutl_sort_by_fields  = NULL;
rec_fex_t  recsel_group_by_fields = NULL;
rec_writer_mode_t recsel_write_mode = REC_WRITER_NORMAL;
char      *recsel_password     = NULL;
bool       recsel_uniq         = false;
size_t     recutl_random       = 0;
char      *recsel_join         = NULL;

/*
 * Command line options management.
 */

enum
{
  COMMON_ARGS,
  RECORD_SELECTION_ARGS,
  PRINT_ARG,
  PRINT_VALUES_ARG,
  PRINT_IN_A_ROW_ARG,
  COLLAPSE_ARG,
  COUNT_ARG,
  DESCRIPTOR_ARG,
  PRINT_SEXPS_ARG,
  SORT_ARG,
#if defined REC_CRYPT_SUPPORT
  PASSWORD_ARG,
#endif
  UNIQ_ARG,
  GROUP_BY_ARG,
  JOIN_ARG
};

static const struct option GNU_longOptions[] =
  {
    COMMON_LONG_ARGS,
    RECORD_SELECTION_LONG_ARGS,
    {"print", required_argument, NULL, PRINT_ARG},
    {"print-values", required_argument, NULL, PRINT_VALUES_ARG},
    {"print-row", required_argument, NULL, PRINT_IN_A_ROW_ARG},
    {"collapse", no_argument, NULL, COLLAPSE_ARG},
    {"count", no_argument, NULL, COUNT_ARG},
    {"include-descriptors", no_argument, NULL, DESCRIPTOR_ARG},
    {"print-sexps", no_argument, NULL, PRINT_SEXPS_ARG},
    {"sort", required_argument, NULL, SORT_ARG},
#if defined REC_CRYPT_SUPPORT
    {"password", required_argument, NULL, PASSWORD_ARG},
#endif
    {"uniq", no_argument, NULL, UNIQ_ARG},
    {"group-by", required_argument, NULL, GROUP_BY_ARG},
    {"join", required_argument, NULL, JOIN_ARG},
    {NULL, 0, NULL, 0}
  };

/*
 * Functions.
 */

void
recutl_print_help (void)
{
  /* TRANSLATORS: --help output, recsel synopsis.
     no-wrap */
  printf (_("\
Usage: recsel [OPTION]... [-t TYPE] [-j FIELD] [-n INDEXES | -e RECORD_EXPR | -q STR | -m NUM] [-c | (-p|-P) FIELD_EXPR] [FILE]...\n"));

  /* TRANSLATORS: --help output, recsel arguments.
     no-wrap */
  fputs(_("\
Select and print rec data.\n"), stdout);

  puts ("");
  /* TRANSLATORS: --help output, recsel arguments.
     no-wrap */
  fputs (_("\
  -d, --include-descriptors           print record descriptors along with the matched\n\
                                        records.\n\
  -C, --collapse                      do not section the result in records with newlines.\n\
  -S, --sort=FIELD,...                sort the output by the specified fields.\n\
  -G, --group-by=FIELD,...            group records by the specified fields.\n\
  -U, --uniq                          remove duplicated fields in the output records.\n"),
         stdout);

#if defined REC_CRYPT_SUPPORT
  /* TRANSLATORS: --help output, encryption related options.
     no-wrap */
  fputs (_("\
  -s, --password=STR                  decrypt confidential fields with the given password.\n"),
         stdout);
#endif
  
  recutl_print_help_common ();

  puts ("");
  recutl_print_help_record_selection ();
  fputs (_("\
  -j, --join=FIELD                    perform an inner join using the specified field.\n"),
         stdout);

  puts ("");
  /* TRANSLATORS: --help output, recsel output options.
     no-wrap */
  fputs (_("\
Output options:\n\
  -p, --print=FIELDS                  comma-separated list of fields to print for each\n\
                                        matching record.\n\
  -P, --print-values=FIELDS           as -p, but print only the values of the selected\n\
                                        fields.\n\
  -R, --print-row=FIELDS              as -P, but separate the values with spaces instead\n\
                                        of newlines.\n\
  -c, --count                         print a count of the matching records instead of\n\
                                        the records themselves.\n"), stdout);

  puts ("");
  /* TRANSLATORS: --help output, recsel special options.
     no-wrap */
  fputs (_("\
Special options:\n\
      --print-sexps                   print the data in sexps instead of rec format.\n"),
         stdout);

  puts ("");
  recutl_print_help_footer ();
}

void
recsel_parse_args (int argc,
                   char **argv)
{
  char c;
  int ret;

  while ((ret = getopt_long (argc,
                             argv,
                             RECORD_SELECTION_SHORT_ARGS
                             ENCRYPTION_SHORT_ARGS
                             "S:Cdcp:P:R:UG:j:",
                             GNU_longOptions,
                             NULL)) != -1)
    {
      c = ret;
      switch (c)
        {
        COMMON_ARGS_CASES
        RECORD_SELECTION_ARGS_CASES
        case DESCRIPTOR_ARG:
        case 'd':
          {
            recsel_descriptors = true;
            break;
          }
        case PRINT_SEXPS_ARG:
          {
            recsel_write_mode = REC_WRITER_SEXP;
            break;
          }
        case UNIQ_ARG:
        case 'U':
          {
            recsel_uniq = true;
            break;
          }
#if defined REC_CRYPT_SUPPORT
        case PASSWORD_ARG:
        case 's':
          {
            if (recsel_password != NULL)
              {
                recutl_fatal (_("more than one password was specified\n"));
              }

            recsel_password = xstrdup (optarg);
            break;
          }
#endif
        case SORT_ARG:
        case 'S':
          {
            if (recutl_sort_by_fields)
              {
                recutl_fatal (_("only one field list can be specified as a sorting criteria.\n"));
              }

            /* Parse the field name.  */

            if (!rec_fex_check (optarg, REC_FEX_CSV))
              {
                recutl_fatal (_("invalid field names in -S.\n"));
              }

            recutl_sort_by_fields = rec_fex_new (optarg, REC_FEX_CSV);
            if (!recutl_sort_by_fields)
              {
                recutl_fatal (_("internal error creating fex.\n"));
              }

            break;
          }
        case JOIN_ARG:
        case 'j':
          {
            if (recsel_join)
              {
                recutl_fatal (_("only one field can be specified as join criteria.\n"));
              }

            if (!rec_field_name_p (optarg))
              {
                recutl_fatal (_("please specify a correct field name to -j|--join.\n"));
              }

            recsel_join = xstrdup (optarg);
            break;
          }
        case GROUP_BY_ARG:
        case 'G':
          {
            if (recsel_group_by_fields)
              {
                recutl_fatal (_("only one field list can be specified as a grouping criteria.\n"));
              }

            /* Parse the field name.  */
            if (!rec_fex_check (optarg, REC_FEX_CSV))
              {
                recutl_fatal (_("invalid field names in -G.\n"));
              }

            recsel_group_by_fields = rec_fex_new (optarg, REC_FEX_CSV);
            if (!recsel_group_by_fields)
              {
                recutl_fatal (_("internal error creating fex.\n"));
              }

            break;
          }
        case PRINT_ARG:
        case PRINT_VALUES_ARG:
        case PRINT_IN_A_ROW_ARG:
        case 'p':
        case 'P':
        case 'R':
          {
            if (recsel_count)
              {
                recutl_fatal (_("cannot specify -[pPR] and also -c.\n"));
              }

            if ((c == 'P') || (c == PRINT_VALUES_ARG))
              {
                recsel_write_mode = REC_WRITER_VALUES;
              }

            if ((c == 'R') || (c == PRINT_IN_A_ROW_ARG))
              {
                recsel_write_mode = REC_WRITER_VALUES_ROW;
              }

            recsel_fex_str = xstrdup (optarg);

            if (!rec_fex_check (recsel_fex_str, REC_FEX_SUBSCRIPTS))
              {
                recutl_fatal (_("invalid list of fields in -%c\n"), c);
              }

            /* Create the field expresion.  */
            recsel_fex = rec_fex_new (recsel_fex_str,
                                      REC_FEX_SUBSCRIPTS);
            if (!recsel_fex)
              {
                recutl_fatal (_("internal error creating the field expression.\n"));
              }

            /* Check that all the functions called in the fex exist.
               Otherwise raise an error.  */
            
            {
              size_t i = 0;
              for (i = 0; i < rec_fex_size (recsel_fex); i++)
                {
                  rec_fex_elem_t elem = rec_fex_get (recsel_fex, i);
                  const char *fname = rec_fex_elem_function_name (elem);

                  if (fname && !rec_aggregate_std_p (fname))
                    {
                      recutl_fatal (_("invalid aggregate function '%s'\n"), fname);
                    }
                }
            }

            break;
          }
        case COLLAPSE_ARG:
        case 'C':
          {
            recsel_collapse = true;
            break;
          }
        case COUNT_ARG:
        case 'c':
          {
            if (recsel_fex_str)
              {
                recutl_fatal (_("cannot specify -c and also -p.\n"));
                exit (EXIT_FAILURE);
              }

            recsel_count = true;
            break;
          }
        default:
          {
            exit (EXIT_FAILURE);
          }

        }
    }

  /* Global checks on the parameters.  */

  if (!recutl_type && recsel_join)
    {
      recutl_fatal (_("joins can only be used when a named record set is selected.\n"));
    }
}

bool
recsel_process_data (rec_db_t db)
{
  int rset_size = 0;
  rec_rset_t rset = NULL;
  rec_writer_t writer = NULL;


#if defined REC_CRYPT_SUPPORT

  /* If recsel was called interactively and with an empty -s, was not
     used then prompt the user for it.  Otherwise use the password
     specified in the command line if any.  */

  if (!recsel_password
      && (recutl_type || (rec_db_size (db) == 1))
      && recutl_interactive ())
    {
      rec_rset_t rset;
      rec_fex_t confidential_fields;

      if (recutl_type)
        {
          rset = rec_db_get_rset_by_type (db, recutl_type);
        }
      else
        {
          rset = rec_db_get_rset (db, 0);
        }

      if (rset)
        {
          confidential_fields = rec_rset_confidential (rset);
          if (rec_fex_size (confidential_fields) > 0)
            {
              recsel_password = recutl_getpass (false);
            }
          
          rec_fex_destroy (confidential_fields);
        }
    }

  /* Note that the password must be at least one character long.  */

  if (recsel_password && (strlen (recsel_password) == 0))
    {
      free (recsel_password);
      recsel_password = NULL;
    }

#endif /* REC_CRYPT_SUPPORT */

  /* If the database contains more than one type of records and the
     user did'nt specify the recutl_type then ask the user to clarify
     the request.  */

  if (!recutl_type && (rec_db_size (db) > 1))
    {
      recutl_fatal (_("several record types found.  Please use -t to specify one.\n"));
    }


  /* Query the database using the criteria specified by the user in
     the command line.  */

  {
    int flags = 0;

    if (recutl_insensitive)
      {
        flags = flags | REC_F_ICASE;
      }

    if (recsel_descriptors)
      {
        flags = flags | REC_F_DESCRIPTOR;
      }

    if (recsel_uniq)
      {
        flags = flags | REC_F_UNIQ;
      }

    rset = rec_db_query (db,
                         recutl_type,
                         recsel_join,
                         recutl_index(),
                         recutl_sex,
                         recutl_quick_str,
                         recutl_random,
                         recsel_fex,
                         recsel_password,
                         recsel_group_by_fields,
                         recutl_sort_by_fields,
                         flags);
    if (!rset)
      recutl_out_of_memory ();
  }

  if (recsel_count)
    {
      /* Write the number of matching records.  */
      
      fprintf (stdout, "%d\n", rec_rset_num_records (rset));
    }
  else
    {
      /* Write the resulting record set to the standard output.  */

      writer = rec_writer_new (stdout);
      rec_writer_set_collapse (writer, recsel_collapse);
      rec_writer_set_skip_comments (writer, true);
      rec_writer_set_mode (writer, recsel_write_mode);
      rec_write_rset (writer, rset);
      rec_writer_destroy (writer);
    }

  rec_rset_destroy (rset);

  return true;
}

int
main (int argc, char *argv[])
{
  int res;
  rec_db_t db;

  res = 0;

  recutl_init ("recsel");

  /* Parse arguments.  */
  recsel_parse_args (argc, argv);

  /* Get the input data.  */
  db = recutl_build_db (argc, argv);
  if (!db)
    {
      res = 1;
      return res;
    }

  /* Process the data.  */
  if (!recsel_process_data (db))
    {
      res = 1;
    }

  rec_db_destroy (db);

  return res;
}

/* End of recsel.c */
